### PR TITLE
Fix `ArrowLooseEvent` not firing for bows

### LINKED ISSUE #1188: ArrowLooseEvent isn't fired when a bow is used
**Minecraft Version:** 1.21

**NeoForge Version:** 21.0.37

**Logs:** No logs, just an event not being fired when a bow is used.

**Steps to Reproduce:**
 1. Subscribe to `ArrowLooseEvent`
 2. Shoot an arrow in game using a bow
 3. Event isn't fired

**Description of issue:**
When a bow is used, `ArrowLooseEvent` isn't fired. When a crossbow is used however, the event *is* fired.

### PR DESCRIPTION
**Pull request by Spinoscythe**
Fixes #1188

This was just a TODO left during porting to 1.20.5

### CODE REVIEW

**Comment by thomasglasser** [comment]
Wouldn't ArrowShotEvent or ArrowFiredEvent be a better name for it 🤔 ArrowLoose sounds weird. Not the point of the PR ik but I thought I'd suggest a rename.

**Comment by thomasglasser** [comment]
Actually it fires for bows no matter what the item is, which may not always be an arrow. BowShotEvent or BowFiredEvent would be better

**Comment by NewJumper** [comment]
or possibly rename it to `ProjectileFireEvent` to better match with the existing `ProjectileImpactEvent`?

**Comment by thomasglasser** [comment]
See that doesn't work bc it doesn't fire for every projectile, it fires from BowItem and CrossbowItem

**Comment by NewJumper** [comment]
oh yea, that's true

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]
Patching the line with the `instanceof ServerLevel` is incorrect. While it prevents the arrow projectile from being spawned, it does not prevent the sound effect, the increment of the statistic, and the projectile item being consumed from the inventory. 

(This should have been tested by adding an always-cancelling listener, which can be mimicked by modifying the `EventHooks` method to always return `-1`.)

The correct place to hook would be just after the `getUseDuration` call, in order to avoid those side-effects I mentioned above, and following the original patch [as seen in 1.20.5](https://github.com/neoforged/NeoForge/blob/8ef74db358eb212361b84b2d671695b983426935/patches/net/minecraft/world/item/BowItem.java.patch#L8-L10).

**Comment by sciwhiz12** [comment]
> Wouldn't ArrowShotEvent or ArrowFiredEvent be a better name for it 🤔

I would assume the name was decided to be the twin for `ArrowNockEvent`, which handles the bow being first drawn. (As to why they are not called `BowDrawnEvent` and `BowFiredEvent`, I don't exactly know.)

**Comment by HenryLoenwind** [comment]
> I would assume the name was decided to be the twin for `ArrowNockEvent`, which handles the bow being first drawn.

That event name is at least 11.5 years old; i.e. it dates back to the time when there was only a bow and a single type of arrow. So, while the Nock event is its twin, the name was chosen because it properly fit back then.

File: patches/net/minecraft/world/item/BowItem.java.patch [hunk 1]
```diff
@@ -1,14 +1,27 @@
...
+-                    if (p_40668_ instanceof ServerLevel serverlevel && !list.isEmpty()) {
+-                        this.shoot(serverlevel, player, player.getUsedItemHand(), p_40667_, list, f * 3.0F, 1.0F, f == 1.0F, null);
++                    if (p_40668_ instanceof ServerLevel serverLevel && !list.isEmpty()) {
++                        this.shoot(serverLevel, player, player.getUsedItemHand(), p_40667_, list, f * 3.0F, 1.0F, f == 1.0F, null);
++
+                     }
+-
+                     p_40668_.playSound(
+-                        null,
++                        player,
```
Comment (Spinoscythe) [inline patches/net/minecraft/world/item/BowItem.java.patch]: Oopsie or not?

Comment (Spinoscythe) [reply patches/net/minecraft/world/item/BowItem.java.patch]: it isn't. phew

Comment (Spinoscythe) [reply patches/net/minecraft/world/item/BowItem.java.patch]: wait it is

File: patches/net/minecraft/world/item/BowItem.java.patch [hunk 2]
```diff
@@ -1,14 +1,24 @@
...
++                i = net.neoforged.neoforge.event.EventHooks.onArrowLoose(itemstack, p_40668_, player, i, !itemstack.isEmpty());
++                if (i < 0) return;
                  float f = getPowerForTime(i);
                  if (!((double)f < 0.1)) {
+                     List<ItemStack> list = draw(p_40667_, itemstack, player);
+-                    if (p_40668_ instanceof ServerLevel serverlevel && !list.isEmpty()) {
+-                        this.shoot(serverlevel, player, player.getUsedItemHand(), p_40667_, list, f * 3.0F, 1.0F, f == 1.0F, null);
++                    if (p_40668_ instanceof ServerLevel serverLevel && !list.isEmpty()) {
++                        this.shoot(serverLevel, player, player.getUsedItemHand(), p_40667_, list, f * 3.0F, 1.0F, f == 1.0F, null);
++
```
Comment (sciwhiz12) [inline patches/net/minecraft/world/item/BowItem.java.patch]: Revert the change to the `ServerLevel` variable name (to the original `serverlevel`; note the capitalization of the middle L).

File: patches/net/minecraft/world/item/BowItem.java.patch [hunk 3]
```diff
@@ -1,14 +1,24 @@
...
+                     }
+-
```
Comment (sciwhiz12) [inline patches/net/minecraft/world/item/BowItem.java.patch]: Add this blank line back.

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]
Seems to work in my testing.

You need to clean up your patches. Re-add back unnecessarily removed blank lines, avoid renaming local variables without reason, etc.

**Review by HenryLoenwind — APPROVED** [review]

**Review by sciwhiz12 — APPROVED** [review]

**Pull request merged**